java/ql/src/experimental/Security/CWE/CWE-089/MyBatisCommonLib.qll

```diff
@@ -0,0 +1,177 @@
+
+private predicate propertiesKey(DataFlow::Node prop, string key) {
+  exists(MethodAccess m |
+    m.getMethod() instanceof PropertiesSetPropertyMethod and
+    key = m.getArgument(0).(CompileTimeConstantExpr).getStringValue() and
+    prop.asExpr() = m.getQualifier()
+  )
+}
+
+/** A data flow configuration tracing flow from ibatis obtaining the variable configuration object to setting the value of the variable. */
+private class PropertiesFlowConfig extends DataFlow2::Configuration {
+  PropertiesFlowConfig() { this = "PropertiesFlowConfig" }
+
+  override predicate isSource(DataFlow::Node src) {
+    exists(MethodAccess ma | ma.getMethod() instanceof IbatisConfigurationGetVariablesMethod |
+      src.asExpr() = ma
+    )
+  }
+
+  override predicate isSink(DataFlow::Node sink) { propertiesKey(sink, _) }
+}
+
+/** Get the key value of Mybatis Configuration Variable. */
+string getAnMybatisConfigurationVariableKey() {
+  exists(PropertiesFlowConfig conf, DataFlow::Node n |
+    propertiesKey(n, result) and
+    conf.hasFlowTo(n)
+  )
+}
+
+/** A reference type that extends a parameterization of `java.util.List`. */
+class ListType extends RefType {
+  ListType() {
+    this.getSourceDeclaration().getASourceSupertype*().hasQualifiedName("java.util", "List")
+  }
+}
+
+/** Gets a call to the MyBatis mapper xml method. */
+MethodAccess getMyBatisMapperXmlMethodAccess(XMLElement xmle) {
+  exists(MyBatisMapperSqlOperation mbmxe |
+    mbmxe.getMapperMethod() = result.getMethod() and
+    (
+      mbmxe.getAChild*() = xmle
+      or
+      exists(MyBatisMapperSql mbms |
+        mbmxe.getInclude().getRefid() = mbms.getId() and
+        mbms.getAChild*() = xmle
+      )
+    )
+  )
+}
+
+/** Gets a call to the MyBatis SQL operation annotation method. */
+MethodAccess getMyBatisSqlOperationAnnotationMethodAccess(IbatisSqlOperationAnnotation isoa) {
+  exists(MyBatisSqlOperationAnnotationMethod msoam |
+    msoam = result.getMethod() and
+    msoam.getAnAnnotation() = isoa
+  )
+}
+
+/** Get the #{...} or ${...} parameters in the Mybatis mapper xml file. */
+private string getAnMybatisXmlSetValue(XMLElement xmle) {
+  result = xmle.getTextValue().trim().regexpFind("(#|\\$)\\{[^\\}]*\\}", _, _)
+}
+
+/** Get the #{...} or ${...} parameters in the Mybatis sql operation annotation value. */
+private string getAMybatisAnnotationSqlValue(IbatisSqlOperationAnnotation isoa) {
+  result = isoa.getSqlValue().trim().regexpFind("(#|\\$)\\{[^\\}]*\\}", _, _)
+}
+
+/** Holds if it is SQL injection of MyBatis xml or MyBatis annotation. */
+predicate isMybatisXmlOrAnnotationSqlInjection(
+  DataFlow::Node node, XMLElement xmle, IbatisSqlOperationAnnotation isoa
+) {
+  exists(MethodAccess ma, string setValue |
+    (
+      ma = getMyBatisMapperXmlMethodAccess(xmle) and
+      setValue = getAnMybatisXmlSetValue(xmle)
+      or
+      ma = getMyBatisSqlOperationAnnotationMethodAccess(isoa) and
+      setValue = getAMybatisAnnotationSqlValue(isoa)
```

Comment:
This is the right idea, but you can't do it with parameters that are only sometimes bound because CodeQL will produce a table that fills in all the values for the unbound parameter.

For example, if I had `class OneToTen extends int { OneToTen() { this = [1,2,3,4,5,6,7,8,9,10] } }` and had a predicate

```
predicate someArgIsFive(OneToTen arg1, OneToTen arg2) { 
  arg1 = 5 or arg2 = 5
}
```

Then CodeQL would produce a table

```
5 | 1
5 | 2
5 | 3
...
5 | 10
1 | 5
2 | 5
3 | 5
...
10 | 5
```

Your `isMybatisXmlOrAnnotationSqlInjection` is like this, except instead of all the numbers 1 to 10 you will get all the elements of the `XMLElement` and `IbatisSqlOperationAnnotation` types. So if there are 100 XML elements in the program named x1 ... x100 of which only x50 matches, and similarly there are 100 IbatisSqlOperationAnnotation instances i1 ... i100 of which only i50 matches, we will produce the result set

```
i1 | x50
i2 | x50
...
i100 | x50
i50 | x1
...
i50 | x100
```

This situation gets worse the more matching XMLElements or IbatisSqlOperationAnnotations we have, as we get closer to producing the cartesian product of the two.

Instead of using these sometimes-bound parameters, you could do:
```
predicate isMybatisXmlOrAnnotationSqlInjection(
  DataFlow::Node node, MethodAccess ma
) {
  exists(string setValue |
    (
      exists(XMLElement xmle | 
        ma = getMyBatisMapperXmlMethodAccess(xmle) and
        setValue = getAnMybatisXmlSetValue(xmle)
      )
      or
      exists(IbatisSqlOperationAnnotation isoa |
        ma = getMyBatisSqlOperationAnnotationMethodAccess(isoa) and
        setValue = getAMybatisAnnotationSqlValue(isoa)
      )
    )
```

Then relate the MethodAccess to the XMLElement or the IbatisSqlOperationAnnotation in the queries that are interested in that particular case.

Reply:
```ql
exists(int i, MyBatisMapperForeach mbmf |
    mbmf = xmle and
    not ma.getMethod().getParameter(i).getAnAnnotation().getType() instanceof TypeParam and
    (
      ma.getMethod().getParameterType(i) instanceof MapType or
      ma.getMethod().getParameterType(i) instanceof ListType or
      ma.getMethod().getParameterType(i) instanceof Array
    ) and
    setValue.matches("%${%}") and
    ma.getArgument(i) = node.asExpr()
 )
```

```mbmf = xmle``` is not included.

Reply:
If that bit is particular to XML, and it appears this is the case, then it should live outside this common function.